rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -688,8 +789,17 @@ extern "C" rmw_node_t * rmw_create_node(
 #endif
 
   dds_qos_t * qos = dds_create_qos();
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    security_options, "rmw_create_node: Unable to create qos", return nullptr);
   std::string user_data = get_node_user_data(name, namespace_);
   dds_qset_userdata(qos, user_data.c_str(), user_data.size());
+
+  if (security_options->enforce_security) {
```

Comment:
This logic is incomplete, it sets the security QoS *only if* `enforce_security` is set.

The expected behavior is to:
If the security files can be found: setup the plugins
If they cannot be found:
- if enforce_security is true -> exit with error message
- else -> create a participant without the security plugins instanciated

This is most likely the cause of the failing tests at https://github.com/ros2/system_tests/pull/408

Reply:
@mikaelarguedas this is how the other middleware behaves, but help me understand if this is the behavior we truly want.

The rcl sets security_options->enforce_security to either RMW_SECURITY_ENFORCEMENT_PERMISSIVE or RMW_SECURITY_ENFORCEMENT_ENFORCE. It's always set to RMW_SECURITY_ENFORCEMENT_PERMISSIVE when ROS_SECURITY_ENABLE is anything other than "true"; that is why it's used to completely bypass security qos here.

With what you propose security will be enabled anytime ROS_SECURITY_ROOT_DIRECTORY exists regardless of ROS_SECURITY_ENABLE and ROS_SECURITY_STRATEGY.  The [rcl only checks for the directory](https://github.com/ros2/rcl/blob/73948da4c50b11d5d133e068b804ce7c2b4f9cb6/rcl/src/rcl/node.c#L289), it (rightfully) does not ensure all six security files exist and are properly formatted. So node creation would fail any time the middleware cannot start with files in the security directory, even when ROS_SECURITY_ENABLE is set to false. 

Similarly ROS_SECURITY_ENABLE=false would not disable security for testing. Instead security would have to be disabled by changing ROS_SECURITY_ROOT_DIRECTORY or moving the security directory on disk.

Reply:
If don't think the described behavior matches how the other middlewares behave, or the desired behavior.
Overall the expected behavior is the one described at: https://design.ros2.org/articles/ros2_dds_security.html

Putting it in a table for ease of read:

| ROS_SECURITY_ENABLE  | ROS_SECURITY_STRATEGY | valid set of security files located under ROS_SECURE_ROOT_DIRECTORY | Expected outcome |
| ----- | ---------- | --- | ----------------------------------------------------------------- |
| false | *          | *   | No security plugins instanciated, creating non-secure participant |
| true  | Enforce    | Yes | creates secure participant                                        |
| true  | Enforce    | No  | fails to create participant                                       |
| true  | Permissive | Yes | creates secure participant                                        |
| true  | Permissive | No  | creates unsecure participant                                      |

Anything not matching that behavior is a bug IMO. 

I agree that the line you are referring to is misleading https://github.com/ros2/rcl/blame/73948da4c50b11d5d133e068b804ce7c2b4f9cb6/rcl/src/rcl/node.c#L316-L317 and in practice serve no purpose and at first glance could be removed completely.
The one that matters in our case is the else branch of that if: https://github.com/ros2/rcl/blame/73948da4c50b11d5d133e068b804ce7c2b4f9cb6/rcl/src/rcl/node.c#L318-L323
If ROS_SECURITY_ENABLE is false: node_security_options.security_root_path is **not** set and this is the variable that tells you if you should try to instantiate the security plugins or not.

Mimicking the logic of the other RMWs could be a good hint of how to implement it here: https://github.com/ros2/rmw_fastrtps/blob/bd5507b61c7285d6083e514ccd49f00b2945baeb/rmw_fastrtps_shared_cpp/src/rmw_node.cpp#L292

Reply:
@mikaelarguedas, thanks for the discussion. It was simple to switch the logic here, but you'll also see changes to `configure_qos_for_security`: all files are checked before any properties are added to qos to enable cyclone dds security.

This is now passing all everything in the `test_security`, that's been a great help and I'd like to see those test routines continue to mature!

Reply:
That's awesome.

> all files are checked before any properties are added to qos to enable cyclone dds security.

Great improvement, we should consider modifying the other rmw implementations to make sure they do the same

> This is now passing all everything in the test_security, that's been a great help and I'd like to see those test routines continue to mature!

:tada: 


I didnt test the new version as it's conflicting since https://github.com/ros2/rmw_cyclonedds/pull/106

Reply:
@SidFaber I gave it a go (with Dashing to verify it supports that, too) and I noticed that it warns
```
[INFO] [rmw_cyclonedds_cpp]: rmw_create_node: Unable to configure security
```
whenever it starts when security either isn't supported or couldn't be configured. That is, when security is:
1. disabled, regardless of whether security support is included in the build;
2. enabled in permissive mode, when security support not included in the build;
3. enabled in permissive mode, when security support is included in the build but something is awry with the configuration.

Case (1) seems wrong (there is no expectation of security so also no expectation of getting warnings about it) and case (3) seems like a good thing to me. I am not quite sure what to make of case (2), I'd say it should be a hard fail but perhaps that's not the intent behind permissive mode (I don't quite get that mode anyway). @mikaelarguedas could you comment on that?

Just for completeness: if security is enabled in enforced mode, it only starts when the security configuration is valid and the implementation supports it.

Apart from the warning not being appropriate in all cases, it looks good to me!

Reply:
> Case (1) seems wrong

:+1:

> case (3) seems like a good thing to me

:+1:

> I am not quite sure what to make of case (2), I'd say it should be a hard fail but perhaps that's not the intent behind permissive mode (I don't quite get that mode anyway)

Originally the "Permissive" mode is to allow someone to launch in a single environment a system with some parts using security and some parts without.
I don't think it's been used much in practice.
In the case of 2, the rmw implementation itself doesnt support security (because not included in the build) so there's no chance of having any part of the system using security. So IMO it should hard fail and provide a clearer error message as of why.

An example is such error message in rmw_fastrtps_cpp https://github.com/ros2/rmw_fastrtps/blob/48b403a211b269ab383e6adfcf89d1b4333b2c39/rmw_fastrtps_shared_cpp/src/participant.cpp#L238-L242

Reply:
How about I just give an info message "Security is enabled" when the cyclone QOS object gets set up properly? It will normally follow the "found security directory" message that comes from rcl. I feel some messaging is important since the rcl message implies that security is enabled which may not be the case.

Reply:
As I find the rcl message by itself pretty noisy when launching systems with many nodes, I'd rather not add an info message *always*. I would prefer a warning when users need to know something is not what they requested (case (3)), and nothing (or debug message) otherwise.

Reply:
@mikaelarguedas, I just removed the info message: there are some tests where you'll get an rcl "found security" message and create a plain text node which is similar to how other middleware behaves. Although not ideal, this is something that would be better solved with logging.